### DISCUSSION TITLE
Single instance breaking OCP upgrade

### DISCUSSION THREAD

**Discussion by aaronmcohen** [cloudnative-pg/cloudnative-pg#517] [Ideas]
_Originally posted by @sxd in https://github.com/cloudnative-pg/cloudnative-pg/issues/486#issuecomment-1197304620_

Hi @aaronmcohen 

A couple of things to notice:
* Please create an issue so we can talk an discuss it there and assign a responsible for this PR
* Fix the linter issue
* Please run the E2E test on this on your repo 

On the other hand, this will basically delete the PDB on every cluster that we create, but we may not want that since this
protects the node from being drained and this needs to be done with during maintenance window and knowing that we
can reuse the PVC, otherwise, this may turn everything into a possible mess, do you have some examples that you can
mention in the issue?

Regards

**Comment by aaronmcohen** [comment]
@sxd I agree a PDB makes sense for an HA deployment. However a single replica is not HA. Adding PDB to a single replica essentially ensures the underlying cluster cannot be upgraded.

**Reply by gbartolini** [reply]
You mean single instance here @aaronmcohen, not single replica, don't you? (meaning instances = 1, one primary, 0 replicas).
Can you please clarify the terminology? Thanks.

**Comment by phisco** [comment]
The PDB is there to prevent unintentional disruption of the Postgres cluster. Moving a single instance could be a risky operation if anything goes wrong with pvcs or anything else could result in the database indefinite unavailability, therefore should not be taken lightly. If possible it's better to always deploy at least 2 instances and if that's not possible we suggest using the maintenance mode.

**Comment by aaronmcohen** [comment]
@phisco I completely agree that if a user wants to prevent unintentional disruption of Postgres they should be running with 2+ instances. I am suggesting that a single replica deployment should not have that expectation and should be used for only POCs. If a Kubernetes/Openshift cluster admin finds a pod preventing a cluster from upgrading they are just going to delete the pod anyway. This of course begs the question what would cause it ti go` wrong with pvcs or anything else could result in the database indefinite unavailability`?

**Comment by phisco** [comment]
There could be valid reasons to deploy single replicas in production, I do agree for POCs it could be good to have the default being not having any pdb, but I think it should be opt-in, leaving the default to be focused toward production use-cases. You can already achieve the same thing by setting both maintenance mode and reuse pvc to true. Maybe we could document it better and make it more clear.

**Comment by aaronmcohen** [comment]
@phisco Why would single replicas in production be valid? A PDB is not going to protect the pod if Kubernetes node fails.

**Reply by phisco** [reply]
Don't know, but I've learned that people often have constraints I can not even imagine. However, the point is, there is already a way to remove the pdb, as I said above, and I think this behavior should be opt-out, not opt-in. It's a design choice, but the above is only my opinion obviously, so if the rest of the team decides differently, I'm happy to get this merged.
CC @sxd @leonardoce @mnencia

**Reply by sxd** [reply]
Yes, there's an option to remove the PDB, it's set the `nodeMaintenanceWindow` and `reusePVC` to `true` this will remove the PDB for the primary instance when there's only one instance, and that piece of code was added because we already had this situation in the past.

If the Kubernetes node fails and you're using a local-storage, yes nothing it's going to protect you from that, but if you're not using local-storage, the data will be protected.

Now the question is, @aaronmcohen when you test this, did you set the `nodeMaintenanceWindow` and `reusePVC` options?
